include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+       *             Jacobian. It is expected that this matrix be of the correct
+       *             size (with dimensions
+       *             <code>n_dependent_variables</code>$\times$<code>n_independent_variables</code>).
+       */
+      static void
+      jacobian(const types::tape_index        active_tape_index,
+               const unsigned int             n_dependent_variables,
+               const std::vector<ScalarType> &independent_variables,
+               FullMatrix<ScalarType> &       jacobian);
+
+      //@}
+    };
+
+
+
+    /**
+     * A prototype driver class for tapeless auto-differentiable numbers.
+     *
+     * It is intended that this class be specialized for the valid
+     * combinations of auto-differentiable numbers and output scalar
+     * number types.
+     *
+     * @tparam ADNumberType A type corresponding to a supported
+     *         auto-differentiable number.
+     * @tparam ScalarType A real or complex floating point number type
+     *         that is the scalar value type used for input to, and output
+     *         from, operations performed with auto-differentiable numbers.
+     * @tparam T An arbitrary type resulting from the application of
+     *         the SFINAE idiom to selectively specialize this class.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename ADNumberType, typename ScalarType, typename T = void>
+    struct TapelessDrivers
+    {
+      // This dummy class definition safely supports compilation
+      // against taped numbers or tapeless number types that have
+      // not yet been implemented.
+
+      /**
+       * @name Configuration
+       */
+      //@{
+
+      /**
+       * In the event that the tapeless mode requires <i>a priori</i> knowledge
+       * of how many directional derivatives might need to be computed, this
+       * function informs the auto-differention library of what this number
+       * is.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       *            that will be used for the entire duration of the
+       *            simulation.
+       *
+       * @warning For ADOL-C tapeless numbers, the value given to
+       * @p n_independent_variables should be the <b>maximum</b> number of
+       * independent variables that will be used for the entire duration of
+       * the simulation. This is important in the context of, for example,
+       * hp-FEM and for multiple constitutive models with a different number of
+       * fields from which a linearization must be computed.
+       */
+      static void
+      initialize_global_environment(const unsigned int n_independent_variables);
+
+      //@}
+
+      /**
+       * @name Drivers for scalar functions
+       */
+      //@{
+
+      /**
+       * Computes the value of the scalar field.
+       *
+       * @param[in] dependent_variables The dependent variables whose values are
+       *            to be extracted.
```

Comment:
I don't think I understand this argument. It's an input argument, so what is extracted from where? And why does one have to input a vector of dependent variables and get a single value out?